### PR TITLE
Compare matched documents to the generated answer

### PR DESCRIPTION
**Pull request by jerpint**
Adding this feature led to a major refactor... I'm very sorry @hbertrand 🥹

### CODE REVIEW

File: buster/validators/base.py [hunk 1]
```diff
@@ -18,54 +18,61 @@ def __init__(self, validator_cfg):
...
             return False
 
         if completion.text == "":
             raise ValueError("Cannot compute embedding of an empty string.")
 
-        response_embedding = get_embedding(
+        unknown_embedding = self.get_embedding(
+            self.unknown_prompt,
+            engine=self.embedding_model,
+        )
```
Comment (hbertrand) [inline buster/validators/base.py:47]: Are you counting on `lru_cache` to cache it?

File: buster/validators/base.py [hunk 2]
```diff
@@ -18,54 +18,61 @@ def __init__(self, validator_cfg):
...
-        score = cosine_similarity(response_embedding, unk_embedding)
-        logger.info(f"UNK score: {score}")
+        unknown_similarity_score = cosine_similarity(answer_embedding, unknown_embedding)
+        logger.info(f"{unknown_similarity_score=}")
 
         # Likely that the answer is meaningful, add the top sources
-        return bool(score < unk_threshold)
+        return bool(unknown_similarity_score < self.unknown_threshold)
+
+    def rerank_docs(self, completion: Completion, matched_documents: pd.DataFrame):
```
Comment (hbertrand) [inline buster/validators/base.py:59]: return typehint

**Review by hbertrand — CHANGES_REQUESTED** [review]

File: buster/busterbot.py [hunk 3]
```diff
@@ -132,67 +67,12 @@ class BusterConfig:
 
 
 class Buster:
-    def __init__(self, cfg: BusterConfig, retriever: Retriever):
-        self._unk_embedding = None
-        self.update_cfg(cfg)
-
+    def __init__(self, retriever: Retriever, completer: Completer, validator):
```
Comment (hbertrand) [inline buster/busterbot.py:70]: typehint validator

File: buster/completers/base.py [hunk 4]
```diff
@@ -60,45 +67,101 @@ def completor(self, value: str) -> None:
...
+            completion_dict["matched_documents"] = pd.DataFrame(completion_dict["matched_documents"]).T
+        else:
+            raise ValueError(f"Unknown type for matched_documents: {type(completion_dict['matched_documents'])}")
+
         # Backwards compatibility
         if "version" not in completion_dict:
             completion_dict["version"] = 1
 
             completion_dict["completor"] = completion_dict["text"]
             del completion_dict["text"]
```
Comment (hbertrand) [inline buster/completers/base.py:105]: I would remove this now that we merged answer and completion

File: buster/completers/base.py [hunk 5]
```diff
@@ -60,45 +67,101 @@ def completor(self, value: str) -> None:
         self._completor = value
 
     def to_json(self) -> Any:
-        to_encode = {"error": self.error, "completor": self.text, "version": self.version}
-        return jsonable_encoder(to_encode)
+        # to_encode = {"error": self.error, "completor": self.text, "version": self.version}
+        # return jsonable_encoder(to_encode)
```
Comment (hbertrand) [inline buster/completers/base.py:71]: remove ?

File: buster/validators/base.py [hunk 6]
```diff
@@ -12,60 +11,79 @@
...
-    def check_sources_used(self, completion: Completion) -> bool:
-        """Check to see if a response is relevant to the chatbot's knowledge or not.
+    def check_answer_relevance(self, answer: str, unknown_prompt: str = None) -> bool:
+        """Check to see if a generated answer is relevant to the chatbot's knowledge or not.
 
         We assume we've prompt-engineered our bot to say a response is unrelated to the context if it isn't relevant.
         Here, we compare the embedding of the response to the embedding of the prompt-engineered "I don't know" embedding.
 
-        set the unk_threshold to 0 to essentially turn off this feature.
+        unk_threshold can be a value between [-1,1]. Usually, 0.85 is a good value.
```
Comment (hbertrand) [inline buster/validators/base.py:32]: Can it really be negative ?

File: buster/validators/base.py [hunk 7]
```diff
@@ -12,60 +11,79 @@
...
+            engine=self.embedding_model,
+        )
+        col = "similarity_to_answer"
+        matched_documents[col] = matched_documents.embedding.apply(
+            lambda x: cosine_similarity(x, answer_embedding) * 100
+        )
+
+        return matched_documents.sort_values(by=col, ascending=False)
+
+    def validate(self, completion: Completion):
```
Comment (hbertrand) [inline buster/validators/base.py:76]: return typehint

File: buster/validators/base.py [hunk 8]
```diff
@@ -12,60 +11,79 @@
...
+        if completion.error:
+            completion.answer_relevant = False
+        elif len(completion.matched_documents) == 0:
+            completion.answer_relevant = False
+        else:
+            completion.answer_relevant = self.check_answer_relevance(completion.text)
+
+        completion.matched_documents = self.rerank_docs(completion.text, completion.matched_documents)
```
Comment (hbertrand) [inline buster/validators/base.py:84]: I thought we wanted this step to be optional ?

**Review by hbertrand — CHANGES_REQUESTED** [review]

File: buster/busterbot.py (hunk 3 above)
Comment (jerpint) [reply buster/busterbot.py:70]: added

File: buster/completers/base.py (hunk 4 above)
Comment (jerpint) [reply buster/completers/base.py:105]: so should we remove the concept of `version` entirely?

File: buster/completers/base.py (hunk 5 above)
Comment (jerpint) [reply buster/completers/base.py:71]: done

File: buster/completers/base.py [hunk 9]
```diff
@@ -60,45 +67,101 @@ def completor(self, value: str) -> None:
...
     @abstractmethod
     def complete(self, prompt) -> str:
         ...
 
-    def generate_response(self, system_prompt, user_input):
+    @abstractmethod
+    def prepare_prompt(self, user_input, matched_documents):
+        ...
+
+    def generate_response(self, user_input: str, matched_documents: pd.DataFrame):
```
[the remark this answers is not in the document]
Comment (jerpint) [reply buster/completers/base.py:131]: fixed, also changed the method name to `get_completion` to be more consistent with the returned typhint

File: buster/examples/gradio_app.py [hunk 10]
```diff
@@ -3,12 +3,24 @@
...
+buster_cfg = cfg.buster_cfg
+retriever: Retriever = SQLiteRetriever(**buster_cfg.retriever_cfg)
+tokenizer = GPTTokenizer(**buster_cfg.tokenizer_cfg)
+completer: Completer = ChatGPTCompleter(
+    documents_formatter=DocumentsFormatter(tokenizer=tokenizer, **buster_cfg.documents_formatter_cfg),
+    prompt_formatter=PromptFormatter(tokenizer=tokenizer, **buster_cfg.prompt_formatter_cfg),
+    **buster_cfg.completion_cfg,
+)
+validator: Validator = Validator(**buster_cfg.validator_cfg)
+buster: Buster = Buster(retriever=retriever, completer=completer, validator=validator)
```
[the remark this answers is not in the document]
Comment (jerpint) [reply buster/examples/gradio_app.py:23]: it's as good as it gets. Every library ive looked at that does similar things (e.g. llama-index) use similar patterns. It's either we make it obvious, and hackable, or we hide everything and make customization much harder and less intuitive. I prefer the former

File: buster/validators/base.py (hunk 6 above)
Comment (jerpint) [reply buster/validators/base.py:32]: i mean in theory cosine distance is in range [-1, 1] so i dont see why not

File: buster/validators/base.py [hunk 11]
```diff
@@ -12,60 +11,79 @@
...
-        score = cosine_similarity(response_embedding, unk_embedding)
-        logger.info(f"UNK score: {score}")
+        unknown_similarity_score = cosine_similarity(answer_embedding, unknown_embedding)
+        logger.info(f"{unknown_similarity_score=}")
 
         # Likely that the answer is meaningful, add the top sources
-        return bool(score < unk_threshold)
+        return bool(unknown_similarity_score < self.unknown_threshold)
+
+    def rerank_docs(self, answer: str, matched_documents: pd.DataFrame):
```
[the remark this answers is not in the document]
Comment (jerpint) [reply buster/validators/base.py:57]: done

File: buster/validators/base.py (hunk 7 above)
Comment (jerpint) [reply buster/validators/base.py:76]: done

File: buster/completers/base.py (hunk 4 above)
Comment (jerpint) [reply buster/completers/base.py:105]: i removed it

File: buster/examples/gradio_app.py [hunk 12]
```diff
@@ -27,16 +27,20 @@ def format_sources(matched_documents: pd.DataFrame) -> str:
     if len(matched_documents) == 0:
         return ""
 
-    sourced_answer_template: str = (
-        """📝 Here are the sources I used to answer your question:<br>""" """{sources}<br><br>""" """{footnote}"""
+    documents_answer_template: str = (
+        """📝 Here are the sources I used to answer your question:<br>""" """{documents}<br><br>""" """{footnote}"""
```
Comment (hbertrand) [inline buster/examples/gradio_app.py:31]: Didn't notice before but the formatting is weird here, why all the """ ?

**Review by hbertrand — APPROVED** [review]

File: buster/examples/gradio_app.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jerpint) [reply buster/examples/gradio_app.py:23]: agreed... i will leave it for yet another round of factoring. I do want to find a better solution for buster_cfg anyway

File: buster/examples/gradio_app.py (hunk 12 above)
Comment (jerpint) [reply buster/examples/gradio_app.py]: looks like it was left overs from a previous refactor, updated to be a more sane single string

**Pull request merged**